### PR TITLE
[AIR][Doc] LightningTrainer Advanced Example

### PR DESCRIPTION
**Pull request by woshiyyya**
## Why are these changes needed?

We only have one toy example for LightningTrainer on MNIST dataset. This PR aims to add an advanced example, which demonstrates how to use LightningTrainer with Ray Data and BatchPredictor. 

- Task: CoLA(Sentence Classification)
- Model: bert-base-uncased
- Resource: 4 GPUs 

Doc build has been failing for a long time, here's the [link to the notebook](https://github.com/ray-project/ray/blob/90c49318158bdf422705181bb313115d6dfaca87/doc/source/train/examples/lightning/lightning_cola_advanced.ipynb).

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [x] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/train/examples/lightning/lightning_cola_advanced.ipynb [hunk 1]
```diff
@@ -0,0 +1,465 @@
...
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "(lightning_advanced_example)=\n",
+    "\n",
+    "# Finetune a BERT Text Classifier with LightningTrainer\n",
+    "\n",
+    ":::{note}\n",
+    "\n",
+    "This is an advanced example for {class}`LightningTrainer <ray.train.lightning.LightningTrainer>`, which demonstrates how to use LightningTrainer with `Ray Dataset` and `Batch Predictor`. \n",
```
Comment (matthewdeng) [inline doc/source/train/examples/lightning/lightning_cola_advanced.ipynb:14]: Recommendation linking to the documentation for the corresponding components:

https://docs.ray.io/en/master/data/dataset.html
https://docs.ray.io/en/master/ray-air/predictors.html

File: doc/source/train/examples/lightning/lightning_cola_advanced.ipynb [hunk 2]
```diff
@@ -0,0 +1,465 @@
...
+    "import ray\n",
+    "import torch\n",
+    "import pytorch_lightning as pl\n",
+    "import torch.nn.functional as F\n",
+    "from torch.utils.data import DataLoader, random_split\n",
+    "from transformers import AutoTokenizer, AutoModelForSequenceClassification\n",
+    "from datasets import load_dataset, load_metric\n",
+    "import numpy as np"
+   ]
+  },
```
Comment (matthewdeng) [inline doc/source/train/examples/lightning/lightning_cola_advanced.ipynb:58]: nit: Suggest adding some text before this, e.g. "Let's start by importing the needed libraries".

File: doc/source/train/examples/lightning/lightning_cola_advanced.ipynb [hunk 3]
```diff
@@ -0,0 +1,465 @@
...
+   "source": [
+    "SMOKE_TEST = True"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [],
+   "source": [
```
Comment (matthewdeng) [inline doc/source/train/examples/lightning/lightning_cola_advanced.ipynb:48]: Can you include instructions for installing dependencies? (`! pip install ...`)

File: doc/source/train/examples/lightning/lightning_cola_advanced.ipynb [hunk 4]
```diff
@@ -0,0 +1,465 @@
...
+    "from datasets import load_dataset, load_metric\n",
+    "import numpy as np"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 1. Pre-process CoLA Dataset\n",
```
Comment (matthewdeng) [inline doc/source/train/examples/lightning/lightning_cola_advanced.ipynb:64]: super nit
```suggestion
    "## 1. Preprocess CoLA Dataset\n",
```

File: doc/source/train/examples/lightning/lightning_cola_advanced.ipynb [hunk 5]
```diff
@@ -0,0 +1,465 @@
...
+    "    encoded_sent = tokenizer(\n",
+    "        batch[\"sentence\"].tolist(),\n",
+    "        max_length=128,\n",
+    "        truncation=True,\n",
+    "        padding=\"max_length\",\n",
+    "        return_tensors=\"pt\",\n",
+    "    )\n",
+    "    batch[\"input_ids\"] = encoded_sent[\"input_ids\"].numpy()\n",
+    "    batch[\"attention_mask\"] = encoded_sent[\"attention_mask\"].numpy()\n",
+    "    batch[\"label\"] = np.array(batch[\"label\"])\n",
```
Comment (matthewdeng) [inline doc/source/train/examples/lightning/lightning_cola_advanced.ipynb:126]: Out of curiosity, what happens if we don't include this conversion?

File: doc/source/train/examples/lightning/lightning_cola_advanced.ipynb [hunk 6]
```diff
@@ -0,0 +1,465 @@
...
+    "    def configure_optimizers(self):\n",
+    "        return torch.optim.AdamW(self.parameters(), lr=self.lr, eps=self.eps)"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 3. Finetune the model with LightningTrainer\n",
```
Comment (matthewdeng) [inline doc/source/train/examples/lightning/lightning_cola_advanced.ipynb:205]: nit: 
```suggestion
    "## 3. Fine-tune the model with LightningTrainer\n",
```

File: doc/source/train/examples/lightning/lightning_cola_advanced.ipynb [hunk 7]
```diff
@@ -0,0 +1,465 @@
...
+    "    for split, ds in ray_datasets.items():\n",
+    "        ray_datasets[split] = ds.random_sample(0.1)"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Train the model with the configuration we specified above. \n",
```
Comment (matthewdeng) [inline doc/source/train/examples/lightning/lightning_cola_advanced.ipynb:275]: I think you can split this into a new header. In particular, the content above this line can be "Configuring your LightningTrainer", and then the content below can be "Fine-tune the model with LightningTrainer".

With this, you can also fill out the section some more by describing `LightningConfigBuilder`, `RunConfig`, and `ScalingConfig` in more detail.

File: doc/source/train/examples/lightning/lightning_cola_advanced.ipynb [hunk 8]
```diff
@@ -0,0 +1,465 @@
...
+    "\n",
+    "To feed data into LightningTrainer, we need to configure the following arguments:\n",
+    "\n",
+    "- datasets: A dictionary of the input Ray datasets, with special keys \"train\" and \"val\".\n",
+    "- datasets_iter_config: The argument list of {meth}`iter_torch_batches() <ray.data.Dataset.iter_torch_batches>`. It defines the way we iterate dataset shards for each worker.\n",
+    "- preprocessor: The preprocessor that will be applied to the input dataset.\n",
+    "\n",
+    ":::{note}\n",
+    "Note that we are using Ray Dataset for data ingestion for faster preprocessing here, but you can also continue to use the native `PyTorch DataLoader` or `LightningDataModule`. See {ref}`this example <lightning_mnist_example>`. \n",
+    "\n",
```
Comment (matthewdeng) [inline doc/source/train/examples/lightning/lightning_cola_advanced.ipynb:285]: Move this under the code block?

**Review by richardliaw — APPROVED** [review]

File: doc/source/train/examples/lightning/lightning_cola_advanced.ipynb [hunk 9]
```diff
@@ -0,0 +1,465 @@
...
+   },
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "Result(\n",
+       "  metrics={'_report_on': 'validation_end', 'train_loss': 0.05989973247051239, 'matthews_correlation': 0.5175218541439164, 'epoch': 4, 'step': 670, 'should_checkpoint': True, 'done': True, 'trial_id': '5ae4c_00000', 'experiment_tag': '0'},\n",
+       "  path='/home/ray/ray_results/ptl-sent-classification/LightningTrainer_5ae4c_00000_0_2023-04-05_12-45-05',\n",
+       "  checkpoint=LightningCheckpoint(local_path=/home/ray/ray_results/ptl-sent-classification/LightningTrainer_5ae4c_00000_0_2023-04-05_12-45-05/checkpoint_000004)\n",
+       ")"
```
Comment (matthewdeng) [inline doc/source/train/examples/lightning/lightning_cola_advanced.ipynb:323]: Can we include the console output of the run?

File: doc/source/train/examples/lightning/lightning_cola_advanced.ipynb [hunk 10]
```diff
@@ -0,0 +1,465 @@
...
+   "outputs": [],
+   "source": [
+    "from ray.train.batch_predictor import BatchPredictor\n",
+    "from ray.train.lightning import LightningCheckpoint, LightningPredictor\n",
+    "\n",
+    "# Use in-memory checkpoint object\n",
+    "checkpoint = result.checkpoint\n",
+    "\n",
+    "# You can also load a checkpoint from disk:\n",
+    "# checkpoint = LightningCheckpoint.from_directory(\"YOUR_CHECKPOINT_DIR\")\n",
```
Comment (matthewdeng) [inline doc/source/train/examples/lightning/lightning_cola_advanced.ipynb:368]: Might be helpful to point out that this the checkpoint directory is what is written in `checkpoint.path`.

File: doc/source/train/examples/lightning/lightning_cola_advanced.ipynb [hunk 11]
```diff
@@ -0,0 +1,465 @@
...
+   ],
+   "source": [
+    "# Internally, BatchPredictor calls forward() method of the LightningModule.\n",
+    "# Convert the logits tensor into labels with argmax.\n",
+    "def argmax(batch):\n",
+    "    batch[\"predictions\"] = batch[\"predictions\"].apply(lambda x: np.argmax(x))\n",
+    "    return batch\n",
+    "\n",
+    "\n",
+    "results = predictions.map_batches(argmax).to_pandas()\n",
```
Comment (matthewdeng) [inline doc/source/train/examples/lightning/lightning_cola_advanced.ipynb:430]: Maybe show how to examine a few of these predictions?

```python
results = predictions.map_batches(argmax)
print(results.take(5))

results_df = results.to_pandas()
...
```

File: doc/source/tune/examples/tune-pytorch-lightning.ipynb [hunk 12]
```diff
@@ -581,6 +581,7 @@
     "## More PyTorch Lightning Examples\n",
     "\n",
     "- {ref}`Use LightningTrainer for Image Classification <lightning_mnist_example>`.\n",
+    "- {ref}`Use LightningTrainer with Ray Data and Batch Predictor <lightning_advanced_example>`\n",
```
Comment (matthewdeng) [inline doc/source/tune/examples/tune-pytorch-lightning.ipynb:584]: Can we also link it at the end of the basic example in the `What’s next?` section?

**Review by matthewdeng — COMMENTED** [review]
Awesome!

File: doc/source/train/examples/lightning/lightning_cola_advanced.ipynb (hunk 2 above)
Comment (woshiyyya) [reply doc/source/train/examples/lightning/lightning_cola_advanced.ipynb:58]: Okey

File: doc/source/train/examples/lightning/lightning_cola_advanced.ipynb (hunk 5 above)
Comment (woshiyyya) [reply doc/source/train/examples/lightning/lightning_cola_advanced.ipynb:126]: Ray dataset is kind of rigid, which only support data format as pandas or numpy. However the output of tokenizer is pytorch tensors, which will throw error if no conversion.

File: doc/source/tune/examples/tune-pytorch-lightning.ipynb (hunk 12 above)
Comment (woshiyyya) [reply doc/source/tune/examples/tune-pytorch-lightning.ipynb:584]: sg. Thanks matt! I'll address these comments in a separate PR.

**Pull request merged**